aiokafka/client.py

```diff
@@ -1,489 +1,335 @@
-                          ReplicaNotAvailableError, check_error)
-from kafka.protocol import KafkaProtocol
+from kafka.common import (KafkaError,
+                          NodeNotReadyError,
+                          KafkaTimeoutError,
+                          UnrecognizedBrokerVersion)
+from kafka.cluster import ClusterMetadata
+from kafka.protocol.metadata import MetadataRequest
+from kafka.protocol.produce import ProduceRequest
 
-from .conn import create_conn
+from aiokafka.conn import create_conn
+from aiokafka import ensure_future
 
 
-__all__ = ['connect', 'AIOKafkaClient']
+__all__ = ['AIOKafkaClient']
 
 
 log = logging.getLogger('aiokafka')
 
 
-@asyncio.coroutine
-def connect(hosts, *, loop=None):
-    if loop is None:
-        loop = asyncio.get_event_loop()
-    client = AIOKafkaClient(hosts, loop=loop)
-    yield from client.load_metadata_for_topics()
-    return client
-
-
 class AIOKafkaClient:
-
-    CLIENT_ID = b"aiokafka-python"
-
-    def __init__(self, hosts, *, client_id=CLIENT_ID, loop):
+    """This class implements interface for interact with Kafka cluster"""
+
+    def __init__(self, *, loop, bootstrap_servers='localhost',
+                 client_id='aiokafka', metadata_max_age_ms=300000,
+                 request_timeout_ms=40000):
+        """Initialize an asynchronous kafka client
+
+        Keyword Arguments:
+            bootstrap_servers: 'host[:port]' string (or list of 'host[:port]'
+                strings) that the consumer should contact to bootstrap initial
+                cluster metadata. This does not have to be the full node list.
+                It just needs to have at least one broker that will respond to
+                Metadata API Request. Default port is 9092. If no servers are
+                specified, will default to localhost:9092.
+            client_id (str): a name for this client. This string is passed in
+                each request to servers and can be used to identify specific
+                server-side log entries that correspond to this client. Also
+                submitted to GroupCoordinator for logging with respect to
+                consumer group administration. Default: 'kafka-python-{ver}'
+            request_timeout_ms (int): Client request timeout in milliseconds.
+                Default: 40000.
+            metadata_max_age_ms (int): The period of time in milliseconds after
+                which we force a refresh of metadata even if we haven't seen
+                any partition leadership changes to proactively discover any
+                new brokers or partitions. Default: 300000
+        """
+        self._bootstrap_servers = bootstrap_servers
         self._client_id = client_id
-        self._hosts = collect_hosts(hosts)
+        self._metadata_max_age_ms = metadata_max_age_ms
+        self._request_timeout_ms = request_timeout_ms
 
+        self.cluster = ClusterMetadata(metadata_max_age_ms=metadata_max_age_ms)
```

Comment:
As I remember ClusterMetadata has locks for threading sync. It can posses overhead... For now we can live with it, but lets consider reimplementing one without sync.